src/dmd/link.d

```diff
@@ -562,59 +562,148 @@ public int runLINK()
+            const(char)* prior;
+            foreach (p; global.params.linkswitches)
             {
-                argv.push("-Xlinker");
-                argv.push(p);
+                // link switches with a prepended "-Xlinker" should be
+                // be pushed together to `linkerArgs`
+
+                if (!isXlinker(p))
+                {
+                    if (!isNullOrEmpty(p) && !flagIsLibraryRelated(p))
+                    {
+                        if (isXlinker(prior))
+                            linkerArgs.push(p);
+                        linkerArgs.push(p);
+                    }
+                }
+
+                prior = p;
             }
         }
+        else
+        {
+            foreach (p; global.params.linkswitches)
+            {
+                if (!isNullOrEmpty(p) && !flagIsLibraryRelated(p))
+                {
+                    linkerArgs.push("-Xlinker");
+                    linkerArgs.push(p);
+                }
+            }
+        }
+
+        // STEP 2
         foreach (p; global.params.libfiles)
         {
             if (FileName.equalsExt(p, "a"))
-                argv.push(p);
+                linkerArgs.push(p);
         }
-        foreach (p; global.params.linkswitches)
+
+        // STEP 3
+        if (global.params.noXlinker) // See https://issues.dlang.org/show_bug.cgi?id=6952
         {
-            if (!p || !p[0] || flagIsLibraryRelated(p))
+            foreach (p; global.params.linkswitches)
             {
-                if (!(p && p[0] == '-' && (p[1] == 'l' || p[1] == 'L')))
+                if (isNullOrEmpty(p) || flagIsLibraryRelated(p))
                 {
-                    // Don't need -Xlinker if switch starts with -l or -L.
-                    // Eliding -Xlinker is significant for -L since it allows our paths
-                    // to take precedence over gcc defaults.
-                    // All other link switches were already added in step 1.
-                    argv.push("-Xlinker");
+                    linkerArgs.push(p);
                 }
-                argv.push(p);
             }
         }
+        else
+        {
+            foreach (p; global.params.linkswitches)
+            {
+                if (isNullOrEmpty(p) || flagIsLibraryRelated(p))
+                {
+                    if (!startsWith(p, "-l") && !startsWith(p, "-L"))
+                    {
+                        // Don't need -Xlinker if switch starts with -l or -L.
+                        // Eliding -Xlinker is significant for -L since it allows our paths
+                        // to take precedence over gcc defaults.
+                        // All other link switches were already added in step 1.
+                        linkerArgs.push("-Xlinker");
+                    }
+                    linkerArgs.push(p);
+                }
+            }
+        }
+
+        // Add accumulated, ordered `linkerArgs` to `argv`
```

Comment:
s/Add/Append/

Reply:
Addressed in https://github.com/dlang/dmd/pull/10341